src/dbnode/client/session.go

```diff
@@ -1419,6 +1421,13 @@ func (s *session) fetchIDsAttempt(
 	// multiple times in case of retries.
 	ids := inputIDs.Duplicate()
 
+	if s.opts.ProtoEnabled() {
+		schema, schemaErr = s.opts.SchemaRegistry().GetSchema(inputNamespace)
```

Comment:
Could we make all of this logic simpler by just creating an object pool that requires you pass a namespace as an argument? Then there is no possibility of future maintainers introducing bugs in this code path.

```golang
type perNSIterPool struct {
    pool encoding.IteratorPool
    registry namespace.SchemaRegistry
}

func (p *perNSIterPool) Get(ns ident.ID) (encoding.Iterator, error) {
    iter := p.pool.Get()
    schema, err := p.registry.GetSchema(ns)
    if err != nil {
        return err
    }
    iter.SetSchema(schema)
    return iter, nil
}
```

Now no one will ever accidentally do this wrong

Reply:
> Could we make all of this logic simpler by just creating an object pool that requires you pass a namespace as an argument? Then there is no possibility of future maintainers introducing bugs in this code path.
> 
> ```go
> type perNSIterPool struct {
>     pool encoding.IteratorPool
>     registry namespace.SchemaRegistry
> }
> 
> func (p *perNSIterPool) Get(ns ident.ID) (encoding.Iterator, error) {
>     iter := p.pool.Get()
>     schema, err := p.registry.GetSchema(ns)
>     if err != nil {
>         return err
>     }
>     iter.SetSchema(schema)
>     return iter, nil
> }
> ```
> 
> Now no one will ever accidentally do this wrong

Regarding robustness, per-namespace pool is safe and backward compatible, in that for proto encoder, if schema is not set, code would crash immediately. That's why I intentionally nil Schema when encoder is returned back to the pool.

Also, pools/options are setup properly during namespace construction and bootstrapper construction, so that other parts of the code do not have to worry about it at all. It is similar to how we do it today, we just delay the pools/options setup from server setup to db construction, rest of the code can continue to depend on options to function.